### PR TITLE
IC-1296 send email when appointment not attended

### PR DESCRIPTION
**Pull request by rgforsyth**
## What does this pull request do?

When recording appointment attendance, if the appointment was not attended, email the pp.

## What is the intent behind these changes?

Email the PP to let them know if an appointment was not attended

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/LocationMapper.kt [hunk 1]
```diff
@@ -19,4 +19,8 @@ class LocationMapper {
   fun mapToCurrentContextPathAsString(path: String, id: UUID): UriComponents {
     return ServletUriComponentsBuilder.fromCurrentContextPath().path(path).buildAndExpand(id)
   }
+
+  fun mapToCurrentContextPathAsAppointment(path: String, id: UUID, sessionsNumber: Int): UriComponents {
+    return ServletUriComponentsBuilder.fromCurrentContextPath().path(path).buildAndExpand(id, id)
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/LocationMapper.kt:24]: there's a typo here, we are expanding `id` twice. a unit test would catch this. 

i also think the method name here is really unhelpful. i wouldn't be able to guess what this method does from the name. same with the method above.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/AppointmentEventPublisher.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ActionPlanAppointment
+import kotlin.reflect.KFunction
+
+enum class AppointmentEventType {
+  NO_ATTENDANCE,
+}
+
+class AppointmentEvent(source: Any, val type: AppointmentEventType, val appointment: ActionPlanAppointment, val detailUrl: String) : ApplicationEvent(source) {
+  override fun toString(): String {
+    return "AppointmentEvent(type=$type, appointment=${appointment.id}, detailUrl='$detailUrl', source=$source)"
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/AppointmentEventPublisher.kt:18]: why the single quotes around the url?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsService.kt [hunk 3]
```diff
@@ -63,9 +65,16 @@ class AppointmentsService(
   ): ActionPlanAppointment {
     val appointment = getActionPlanAppointmentOrThrowException(actionPlanId, sessionNumber)
     setAttendanceFields(appointment, attended, additionalInformation)
+    if (attended == Attended.NO) {
+      notifyNoAttendance(appointment)
+    }
     return actionPlanAppointmentRepository.save(appointment)
   }
 
+  private fun notifyNoAttendance(appointment: ActionPlanAppointment) {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsService.kt:74]: does this need to be in its own method?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt [hunk 4]
```diff
@@ -49,6 +51,35 @@ class NotifyActionPlanService(
...
+  @Value("\${interventions-ui.locations.appointment-not-attended}") private val interventionsUIAppointmentNotAttendedLocation: String,
+  private val emailSender: EmailSender,
+  private val hmppsAuthService: HMPPSAuthService,
+) : ApplicationListener<AppointmentEvent>, NotifyService {
+
+  @AsyncEventExceptionHandling
+  override fun onApplicationEvent(event: AppointmentEvent) {
+    when (event.type) {
+      AppointmentEventType.NO_ATTENDANCE -> {
+        val userDetails = hmppsAuthService.getUserDetail(event.appointment.actionPlan.referral.createdBy)
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt:67]: i'm sure we talked about this but i can't remember the outcome - do we want the notification to go to the PP who created the referral, or the one who sent it? 

either way, we should ensure we are consistent with these types of notifications. could we add a helper method to the referral for 'get responsible pp' or something?

Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt:67]: nitpick: i would also rename this variable so it's clear who's user details these are.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt [hunk 5]
```diff
@@ -49,6 +51,35 @@ class NotifyActionPlanService(
...
+        val location = generateResourceUrl(interventionsUIBaseURL, interventionsUIAppointmentNotAttendedLocation, event.appointment.id)
+        emailSender.sendEmail(
+          appointmentNotAttendedTemplateID,
+          userDetails.email,
+          mapOf(
+            "ppFirstName" to userDetails.firstName,
+            "referralNumber" to event.appointment.actionPlan.referral.referenceNumber!!,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt:74]: we should probably update this var name in the template so it's clear it's expecting the reference number. 'referralNumber' is ambiguous - it could be the referral id too.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt [hunk 6]
```diff
@@ -49,6 +51,35 @@ class NotifyActionPlanService(
...
+            "deliusAbsence" to location.toString(),
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt:75]: same with this variable - it's not clear that it's expecting a link

File: src/main/resources/application.yml [hunk 7]
```diff
@@ -76,11 +76,13 @@ notify:
     referral-sent: "336f2158-d130-4c63-a59b-da12df15a5a3"
     referral-assigned: "f0fe2fce-df0f-46c0-839a-822037d39c28"
     action-plan-submitted: "83743b43-9abb-4ffb-913d-dc583ab22dbf"
+    appointment-not-attended: "21b14c86-30d2-4bf2-b355-8f3deb62ef69"
 
 interventions-ui:
   locations:
     sent-referral: "/service-provider/referrals/{id}"
     submit-action-plan: "/action-plan/{id}/toBeSpecified"
+    appointment-not-attended: "/appointment/{id}/toBeSpecified"
```
Comment (tomsmyers) [inline src/main/resources/application.yml:85]: not a blocker - is it better to have these links point to the pp dashboard, so at least they actually work, rather than to a 404 page?

**Review by tomsmyers — CHANGES_REQUESTED** [review]
nice work

one change required (URL params) and some smaller things to talk through

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsService.kt (hunk 3 above)
Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsService.kt:74]: Perhaps the check for non-attendance should be in the method and the method called notifyIfNonAttendance which means the main method just orchestrates

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/LocationMapper.kt (hunk 1 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/LocationMapper.kt:24]: this has been addressed in a separate change so is no longer an issue

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/AppointmentEventPublisher.kt (hunk 2 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/AppointmentEventPublisher.kt:18]: leaving for now since it matches the other event string

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsService.kt (hunk 3 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentsService.kt:74]: 👍  done

File: src/main/resources/application.yml (hunk 7 above)
Comment (tomsmyers) [reply src/main/resources/application.yml]: done 

https://dsdmoj.atlassian.net/browse/IC-1482
https://dsdmoj.atlassian.net/browse/IC-1483

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/LocationMapper.kt [hunk 8]
```diff
@@ -1,12 +1,22 @@
...
+import kotlin.reflect.KFunction
+import kotlin.reflect.full.findAnnotation
 
 @Component
 class LocationMapper {
   fun expandPathToCurrentRequestBaseUrl(path: String, vararg uriVariableValues: Any): URI {
     return ServletUriComponentsBuilder.fromCurrentRequestUri().path(path).buildAndExpand(*uriVariableValues).toUri()
   }
+  
+  fun getPathFromControllerMethod(method: KFunction<*>): String {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/LocationMapper.kt:17]: this is split into its own commit. please review that in isolation to see this change.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt [hunk 9]
```diff
@@ -49,6 +51,35 @@ class NotifyActionPlanService(
...
+    when (event.type) {
+      AppointmentEventType.NO_ATTENDANCE -> {
+        val ppDetails = hmppsAuthService.getUserDetail(event.appointment.actionPlan.referral.getResponsibleProbationPractitioner())
+        val location = generateResourceUrl(interventionsUIBaseURL, interventionsUIAppointmentNotAttendedLocation, event.appointment.id)
+        emailSender.sendEmail(
+          appointmentNotAttendedTemplateID,
+          ppDetails.email,
+          mapOf(
+            "ppFirstName" to ppDetails.firstName,
+            "referenceNumber" to event.appointment.actionPlan.referral.referenceNumber!!,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/NotifyService.kt:74]: i have updated these var names in the template

File: src/main/resources/application.yml [hunk 10]
```diff
@@ -76,11 +76,13 @@ notify:
     referral-sent: "336f2158-d130-4c63-a59b-da12df15a5a3"
     referral-assigned: "f0fe2fce-df0f-46c0-839a-822037d39c28"
     action-plan-submitted: "83743b43-9abb-4ffb-913d-dc583ab22dbf"
+    appointment-not-attended: "21b14c86-30d2-4bf2-b355-8f3deb62ef69"
 
 interventions-ui:
   locations:
     sent-referral: "/service-provider/referrals/{id}/details"
```
Comment (mahalrmoj) [inline src/main/resources/application.yml:83]: Just a thought but should this one be progress to rather than service provider? It is used in the NSI/Referrals url in delius

**Review by mahalrmoj — APPROVED** [review]

File: src/main/resources/application.yml [hunk 11]
```diff
@@ -76,11 +76,13 @@ notify:
...
-    submit-action-plan: "/action-plan/{id}/toBeSpecified"
+    submit-action-plan: "/service-provider/referrals/{id}/progress" // fixme
```
[the remark this answers is not in the document]
Comment (tomsmyers) [reply src/main/resources/application.yml:84]: yeh right now there is no pp facing screen with this info. but with the current implementation the pp can actually load this page. both ways are not ideal, but i thought perhaps this way was better than an error?

**Review by tomsmyers — APPROVED** [review]
